Fix debugee to inherit original environment variables ##debug

abcSup opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Duplicate and save the original `environ` and use it for `execvpe` later in `io_debug` so the child process will not inherit the `environ` of `r2`. 

**Test plan**
Run `env -i /usr/local/bin/r2 -c dc -d /usr/bin/env` and check for empty outputs. However, afaik there is no way to run command like that in the test suite. This test should be done in debug mode, so a custom binary is required to check for environment variables set by `r2` and one of which is in `PATH`. But, this way will make the test depends on the binary and any changes in future will require the binary to be replaced to make the test valid which is too troublesome. I would be glad to receive any suggestion on this.

**Closing issues**

closes #16359

abcSup commented:
> This pull request **introduces 1 alert** when merging [9114064](https://github.com/radareorg/radare2/commit/9114064bc5b2475a15cf9d256a9a55d5258a02c6) into [8540156](https://github.com/radareorg/radare2/commit/8540156f7397372968064157d7ad1473457cfa31) - [view on LGTM.com](https://lgtm.com/projects/g/radareorg/radare2/rev/pr-a717969132dddad6854aa970003f6fa3bbd6482e)
> 
> **new alerts:**
> 
> * 1 for Implicit function declaration

I found out that I should include the `#define _GNU_SOURCE` feature test to use `execvpe`. I just added that at the top of `io_debug.c`. Let me know if that works. 

Although the current `io_debug` does not support FreeBSD, I included a check for it because FreeBSD's equivalent of `execvpe` is `exect` to satisfy the FreeBSD build test.

trufae commented on libr/util/sys.c:
```diff
@@ -1199,10 +1199,32 @@ R_API char **r_sys_get_environ (void) {
+			;
+		count = i;
+
+		origenv = malloc (sizeof (char*) * (count + 1));
```
use calloc and avoid the multiplication. Also this global allocation is never free'd and this can be a problem as long as nothing ensures this is called only once and there's no way to deallocate it.

Also, env is writable. and its linearly allocated in memory, we can just do a single allocation for the whole buffer and then another allocation for the pointer storage. instead of having one strdup per-env. But this can result on some other problems we probably dont want to handle.

What are your thoughs on this @ret2libc ?

ret2libc replied:
Use `R_NEWS`

ret2libc replied:
Although in theory what @trufae suggested is probably the best thing memory-wise, I would stick with this simpler solution for the beginning, unless we realize those additional strdup are really a problem.

ret2libc commented on libr/util/sys.c:
```diff
@@ -26,6 +26,7 @@
 #include <r_lib.h>
 
 static char** env = NULL;
+static char** origenv = NULL;
```
I don't particularly like this global `origenv`... even though there was already a global `env` as well :(

trufae replied:
Me neither but i don't see any other option here

ret2libc commented on libr/util/sys.c:
```diff
@@ -1199,10 +1199,32 @@ R_API char **r_sys_get_environ (void) {
+	if (env) {
+		size_t count, i;
+
+		for (i = 0; env[i]; i++)
```
just use `count` directly

ret2libc commented on libr/io/p/io_debug.c:
```diff
@@ -1,5 +1,6 @@
 /* radare - LGPL - Copyright 2007-2019 - pancake */
 
+#define _GNU_SOURCE
```
I'm not sure we want to depend on this. @trufae ? This could break on some non-gnu platforms? I'm not sure what they are, but i think there are some we want to support.

trufae replied:
i assume this is needed for `execvpe` which is not available on bsd or mac for example. isnt execve enough?

abcSup commented:
Thanks for the feedbacks! I will get back to this soon after finishing another PR for testing remote debugging.

trufae commented:
What about saving the startup environment as a rarun profile? and add a command to print and edit it with cfg.editor? this way we don't need to hold that environment differently than when the user specifies a rarun profile and will let us modify the execution environment easily for the next run

abcSup commented:
> What about saving the startup environment as a rarun profile? and add a command to print and edit it with cfg.editor? this way we don't need to hold that environment differently than when the user specifies a rarun profile and will let us modify the execution environment easily for the next run

That sounds better and cleaner. I will check out rarun profile and implement it with that then. Thanks!

trufae commented:
Great! I would love to see that!

XVilka commented:
@abcSup are there any updates on this PR? It would be awesome to include the fix in the imminent 4.5.0 release.

abcSup commented:
> @abcSup are there any updates on this PR? It would be awesome to include the fix in the imminent 4.5.0 release.

Ok, I will finish this now to be included in the new release. I will also try to test this on a Windows and Mac to make sure it work for most platforms.

abcSup commented:
**Changes**
- Retrieve original environment and save it into `io->envprofile` in rarun2 profile format
- In Linux & Mac, always spawn new process with rarun2 profile from either user-defined profile file `dbg.profile` or the default rarun2 profile in `io->envprofile`.
- Changed `env` command to display the default rarun2 profile saved in `io->envprofile`
- Added `env!` command to edit default rarun2 profile with `cfg.editor`

Let me know:
- If I should change the command `env` to something more intuitive (maybe something starts with `o` or `d`).
- If I should change `r2` to setup those environment variables on Mac & Windows (which is the current implementation).
  - On Windows, 7th argument is `NULL`
https://github.com/radareorg/radare2/blob/4b029876660c5cac00d79066e27e0539895d3946/libr/io/p/io_debug.c#L106-L108
  - On Mac, last argument is `NULL`
https://github.com/radareorg/radare2/blob/4b029876660c5cac00d79066e27e0539895d3946/libr/io/p/io_debug.c#L396

abcSup commented:
I keep the original `env` command and introduce two new commands `doe` and `doe!` to show and edit the rarun2 startup profile.

trufae commented on libr/core/cmd_debug.c:
```diff
@@ -5383,6 +5385,24 @@ static int cmd_debug(void *data, const char *input) {
+		case 'e': // "doe"
+			switch (input[2]) {
+			case '\0': // "doe"
+				r_cons_println (core->io->envprofile);
```
envprofile can be null, add a chk

yossizap commented:
>If I should change r2 to setup those environment variables on Mac & Windows native debugger (the current implementation does not include environment variables when spawning a new process). 

That would be nice for windows but it can be done in a future PR since custom environment variables are very rarely used, the created process uses system default which is fine.

trufae commented:
@thestr4ng3r can you review this pr?

XVilka commented:
Can we merge this?

abcSup commented:
I just noticed and fixed a logic error in `r_run_parse`, waiting for review.
